capsules/src/rf233.rs

```diff
@@ -0,0 +1,885 @@
+            receiving: Cell::new(false),
+            spi_busy: Cell::new(false),
+            state: Cell::new(InternalState::START),
+            interrupt_handling: Cell::new(false),
+            interrupt_pending: Cell::new(false),
+            tx_buf: TakeCell::empty(),
+            rx_buf: TakeCell::empty(),
+            tx_len: Cell::new(0),
+            tx_client: Cell::new(None),
+            rx_client: Cell::new(None),
+            addr: Cell::new(0),
+            pan: Cell::new(0),
+            seq: Cell::new(0),
+            spi_rx: TakeCell::empty(),
+            spi_tx: TakeCell::empty(),
+            spi_buf: TakeCell::empty(),
+        }
+    }
+
+    fn handle_interrupt(&self) {
+        // Because the first thing we do on handling an interrupt is
+        // read the IRQ status, we defer handling the state transition
+        // to the SPI handler
+        if self.spi_busy.get() == false {
+            self.interrupt_handling.set(true);
+            self.register_read(RF233Register::IRQ_STATUS);
+        } else {
+            self.interrupt_pending.set(true);
+        }
+    }
+
+    #[allow(dead_code)]
+    fn register_write(&self, reg: RF233Register, val: u8) -> ReturnCode {
+
+        if (self.spi_busy.get() || self.spi_tx.is_none() || self.spi_rx.is_none()) {
+            return ReturnCode::EBUSY;
+        }
+        let wbuf = self.spi_tx.take().unwrap();
+        let rbuf = self.spi_rx.take().unwrap();
+        wbuf[0] = (reg as u8) | RF233BusCommand::REGISTER_WRITE as u8;
+        wbuf[1] = val;
+        self.spi.read_write_bytes(wbuf, Some(rbuf), 2);
+        self.spi_busy.set(true);
+
+        ReturnCode::SUCCESS
+    }
+
+    fn register_read(&self, reg: RF233Register) -> ReturnCode {
+
+        if (self.spi_busy.get() || self.spi_tx.is_none() || self.spi_rx.is_none()) {
+            return ReturnCode::EBUSY;
+        }
+
+        let wbuf = self.spi_tx.take().unwrap();
+        let rbuf = self.spi_rx.take().unwrap();
+        wbuf[0] = (reg as u8) | RF233BusCommand::REGISTER_READ as u8;
+        wbuf[1] = 0;
+        self.spi.read_write_bytes(wbuf, Some(rbuf), 2);
+        self.spi_busy.set(true);
+
+        ReturnCode::SUCCESS
+    }
+
+    fn frame_write(&self, buf: &'static mut [u8], buf_len: u8) -> ReturnCode {
+        if self.spi_busy.get() {
+            return ReturnCode::EBUSY;
+        }
+
+        let op_len = (buf_len + 1) as usize;
+        buf[0] = RF233BusCommand::FRAME_WRITE as u8;
+        self.spi.read_write_bytes(buf, self.spi_buf.take(), op_len);
+        self.spi_busy.set(true);
+        ReturnCode::SUCCESS
+    }
+
+    // A future, correct version will use buf and properly handle buffers.
+    #[allow(unused_variables)]
+    fn frame_read(&self, buf: &'static mut [u8], buf_len: u8) -> ReturnCode {
+        if self.spi_busy.get() {
+            return ReturnCode::EBUSY;
+        }
+        let op_len = (buf_len + 1) as usize;
+        let wbuf = self.spi_buf.take().unwrap();
```

Comment:
This one is tricky. If spi_buf is empty, then there are a lot of problems going on; backing the stack out into a reasonable state would be... hard. I'd personally rather just panic.